### ISSUE TITLE
Generated serial number for workouts

### ISSUE THREAD

**Issue by surbeym** [jpickup/GarminTools#10]
Could we generate or have a user provide the serial number for generated workouts. Right now when generating the workout's serial_number is always set to 1. 

This prevents use of using the schedule.fit as all the workout in it are set to 1. 

The issue this creates is all the workouts are on my Garmin device and the schedule.fit is but the calendar is empty

Ideally with two workouts the serial numbers would be unique, so that in the schedule.fit they would be different,

Current fit file converted to csv
Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067568026",,scheduled_time,"1099983600",,serial_number,"1",,completed,"0",,
Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100070000",,serial_number,"1",,completed,"0",,
Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100156400",,serial_number,"1",,completed,"0",,

...

Expected
Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067568026",,scheduled_time,"1099983600",,serial_number,"18711",,completed,"0",,
Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100070000",,serial_number,"4518",,completed,"0",,
Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100156400",,serial_number,"139147",,completed,"0",,

These serials would match to the serials of the workout,fit files that were generated.

**Comment by jpickup** [comment]
Sure, I'll fix this, most likely generating a serial rather than a user input. I suspect that this must only be an issue on newer Garmins as my, now rather old, Fenix 3 doesn't seem to have a problem with this.

**Comment by jpickup** [comment]
The Garmin docs [here](https://developer.garmin.com/fit/file-types/) seem to indicate that the serial number would normally be the serial number of the watch, well for FIT files generated by the watch anyway. For FIT files generated externally that doesn't make sense, hence my use of a default value of 1. I think it would be better if I could reproduce the issue you are experiencing before making changes. Could you describe the steps to create this? Perhaps also include a sample XLS file.

To try and eliminate my Fenix 3 being an old watch that doesn't care about something that newer watches have a problem with I just tried this on a Fenix 7 and that seems happy too, but perhaps I'm doing something differently to you - hence wanting to know steps to follow.

One further question: are you running the app on Windows or MacOS? Only reason I ask is that it's much easier for me to build for Mac so if I do create a new version for you to test then I'll be able to turn it around much faster if I only create the Mac bundle during the test phase, i.e. I'd leave building the Windows installer until testing was complete/successful

**Comment by Abasz** [comment]
> Could we generate or have a user provide the serial number for generated workouts. Right now when generating the workout's serial_number is always set to 1.
> 
> This prevents use of using the schedule.fit as all the workout in it are set to 1.
> 
> The issue this creates is all the workouts are on my Garmin device and the schedule.fit is but the calendar is empty
> 
> Ideally with two workouts the serial numbers would be unique, so that in the schedule.fit they would be different,
> 
> Current fit file converted to csv Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067568026",,scheduled_time,"1099983600",,serial_number,"1",,completed,"0",, Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100070000",,serial_number,"1",,completed,"0",, Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100156400",,serial_number,"1",,completed,"0",,
> 
> ...
> 
> Expected Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067568026",,scheduled_time,"1099983600",,serial_number,"18711",,completed,"0",, Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100070000",,serial_number,"4518",,completed,"0",, Data,0,schedule,manufacturer,"1",,garmin_product,"65534",,type,"0",,time_created,"1067567993",,scheduled_time,"1100156400",,serial_number,"139147",,completed,"0",,
> 
> These serials would match to the serials of the workout,fit files that were generated.

These timestamps are rather strange: 

Garmin does not use EPOCH for timestamps but rather their own (no comment)... 

So if the scheduled time timestamps are already converted to EPOCH, then these are in 2004 :) If not then in 2024. That could be one of the reasons not showing up. I mean they show up but for 10 November 2024 (I suppose you were trying to schedule for 10 November 2023).

Serial number cannot be an issue. I have worked a lot with the FIT file SDK and that is not really a relevant field for these things. I am quite convinced that the date is wrong. If possible, can you please check your calendar for 2024 (I am not sure if this is possible though:))

**Comment by jpickup** [comment]
I'm fairly confident that the date conversions work as scheduled workouts do appear on my watch for the correct date in the future. One thing to point out: the watch, or at least the Fenix 3, has a limit of 30 workouts in a schedule so you'll only see the next 30, none any beyond that. I don't know what the limits are on other watches. For that reason the schedule file only includes future workouts, never those in the past, as this was the only way I could ensure I had the next few weeks on the watch and just had to re-upload every month or so.

Does that explain the issue? If not, can you send me the XLS you are using?

**Comment by surbeym** [comment]
I'll be happy to provide more data when back home tonight!

**Comment by Abasz** [comment]
> I'm fairly confident that the date conversions work as scheduled workouts do appear on my watch for the correct date in the future. One thing to point out: the watch, or at least the Fenix 3, has a limit of 30 workouts in a schedule so you'll only see the next 30, none any beyond that. I don't know what the limits are on other watches. For that reason the schedule file only includes future workouts, never those in the past, as this was the only way I could ensure I had the next few weeks on the watch and just had to re-upload every month or so.
> 
> Does that explain the issue? If not, can you send me the XLS you are using?

Sorry if my comment was missunderstandable, I was not suggesting that the conversion is wrong in the code, rather that the entered date is. Time created (1067568026) seems correct: 30 October 2023, but scheduled time (1099983600) is 8 November 2024. If this was intentional than it is clearly not showing as it is out of bound for the watch.

**Comment by Abasz** [comment]
> I'm fairly confident that the date conversions work as scheduled workouts do appear on my watch for the correct date in the future. One thing to point out: the watch, or at least the Fenix 3, has a limit of 30 workouts in a schedule so you'll only see the next 30, none any beyond that. I don't know what the limits are on other watches. For that reason the schedule file only includes future workouts, never those in the past, as this was the only way I could ensure I had the next few weeks on the watch and just had to re-upload every month or so.
> 
> Does that explain the issue? If not, can you send me the XLS you are using?

Sorry if my comment was missunderstandable, I was not suggesting that the conversion is wrong in the code, rather that the entered date is. Time created (1067568026) seems correct: 30 October 2023, but scheduled time (1099983600) is 8 November 2024. If this was intentional than it is clearly not showing as it is out of bound for the watch.

**Comment by jpickup** [comment]
@Abasz - I understand you now, sorry. I guess the answer will lie in the XLS from @surbeym. 

One thing you say "as it is out of bound for the watch" is interesting as I always assumed that I wasn't seeing scheduled workouts out in the future because I was trying to load too many (usually 5 per week over a 5 month period) and I concluded, perhaps incorrectly, that there was a limit of 30 future scheduled workouts but it may well be that there is a hard limit on how far in the future the workout is scheduled for rather than how many. Do you know if any of this is documented by Garmin?

**Comment by surbeym** [comment]
Hello I have created a new repo with the files https://github.com/surbeym/issue10, could not think of a better way.

I am using Fenix 6X on windows

**Comment by Abasz** [comment]
> Hello I have created a new repo with the files https://github.com/surbeym/issue10, could not think of a better way.
> 
> I am using Fenix 6X on windows

You are scheduling for 2024. That is not going to work. 

My experience is that the watch considers only 30 days (1 month) ahead. Though I dont have direct experience with adding directly a schedule.fit as I use the GC calendar for scheduling that gets synced to the watch, it only syncs 30 days worth of training. 

Now it is possible that if you schedule ahead of the 30 days once a workout gets within bound of that time frame it shows up, but I dont have specific experience on this.

**Comment by Abasz** [comment]
So after writing the above, and reading through the whole thing again I realized that I might have mislead you when I said that serial number is irelvant. In general on activity and other files it has not much relevance. But actually for schedules the serial number is used to identify the workout.

When you look at the source code in the FIT SDK in the SchedlueMesg.cs file (yes its C# but its in Java and all others too :)) there is a comment.

```cs
 ///<summary>
        /// Retrieves the SerialNumber field
        /// Comment: Corresponds to file_id of scheduled workout / course.</summary>
        /// <returns>Returns nullable uint representing the SerialNumber field</returns>
        public uint? GetSerialNumber()
        {
            Object val = GetFieldValue(2, 0, Fit.SubfieldIndexMainField);
            if(val == null)
            {
                return null;
            }

            return (Convert.ToUInt32(val));
            
        }
```

Now the strange part is that actually this comment is included in every file_id field which I have not noticed previously

Also if you look at the workout example in the FIT SDK documentation https://developer.garmin.com/fit/cookbook/encoding-workout-files/ They set a date as serial number 

```cs
    fileIdMesg.SetSerialNumber(timeCreated.GetTimeStamp());
```

Also in the SDK examples cookbook for wokrout there is a comment in the C# code:

```cs
 static void CreateWorkout(WorkoutMesg workoutMesg, List<WorkoutStepMesg> workoutSteps)
        {
            // The combination of file type, manufacturer id, product id, and serial number should be unique.
            // When available, a non-random serial number should be used.
            Dynastream.Fit.File fileType = Dynastream.Fit.File.Workout;
            ushort manufacturerId = Manufacturer.Development;
            ushort productId = 0;
            Random random = new Random();
            uint serialNumber = (uint)random.Next();
```

Now of course in the above they say non-random and then they generate a random number which seems contradictory but the point is the fact that it needs to be unique. 

I also attached a schedlue.fit and a schedlued workout that has been shcedlued natively from GC in json format
[Schedule.json](https://github.com/jpickup/GarminTools/files/13225179/Schedule.json)
[2023-11-23.json](https://github.com/jpickup/GarminTools/files/13225180/2023-11-23.json)

**Comment by Abasz** [comment]
@surbeym can you please confirm if you are copying the new files to the NewFiles folder on the watch or directly to the schedlue and workouts folder? 

On newer watches files should be copied to NewFiles and when disconnected from computer the watches installs them automatically. I was not able to find this instructions for Fenix or FR255 but on the support pages this is the instruction for Edge devices, and this works on my watch too.

**Comment by Abasz** [comment]
> @Abasz - I understand you now, sorry. I guess the answer will lie in the XLS from @surbeym.
> 
> One thing you say "as it is out of bound for the watch" is interesting as I always assumed that I wasn't seeing scheduled workouts out in the future because I was trying to load too many (usually 5 per week over a 5 month period) and I concluded, perhaps incorrectly, that there was a limit of 30 future scheduled workouts but it may well be that there is a hard limit on how far in the future the workout is scheduled for rather than how many. Do you know if any of this is documented by Garmin?

So this is a bit strange actually, I never tested the maximum number of scheduleable calendar workouts, but I was referring to the docs: https://support.garmin.com/en-US/?faq=aEyhNQKOCFAytMFnCvdAM7 as well as to my experience where syncing from GC calendar to the watch was limited to 30 days ahead (i.e. GC only syncs a month ahead).

Now I have tested @surbeym xls file and copied all 113 files to the NewFiles folder, after disconnecting the watch I get a warning that the _"maximum number of workouts exceeded"_ (that is ok and expected :)) but strangely I can see 30 workouts scheduled for 2024 (though the year is not showing, but I scheduled from the 1 November and it does not show as today so year is seemingly adhered :) which was not my expectation. Now in the workouts (non-scheduled folder) I can see 25.

**So basically my previously thought bounds (i.e. the 1 month time limit) only applies for syncing scheduled workouts from GC calendar and on the manual side its 30 scheduled workout that applies.**

**Comment by surbeym** [comment]
> @surbeym can you please confirm if you are copying the new files to the NewFiles folder on the watch or directly to the schedlue and workouts folder?

I directly dropped them in 
Schedule.fit file -> This PC\fenix 6X Pro Solar\Primary\GARMIN\Schedule,
Workouts -> This PC\fenix 6X Pro Solar\Primary\GARMIN\Workouts\Schedule

**Comment by Abasz** [comment]
> > @surbeym can you please confirm if you are copying the new files to the NewFiles folder on the watch or directly to the schedlue and workouts folder?
> 
> I directly dropped them in Schedule.fit file -> This PC\fenix 6X Pro Solar\Primary\GARMIN\Schedule, Workouts -> This PC\fenix 6X Pro Solar\Primary\GARMIN\Workouts\Schedule

Try adding them to the NewFiles folder 

> Curious how did you get them in json format instead of csv?

I wrote my own parser: [FitFileConverter](https://github.com/Abasz/FitFileConverter)

**Comment by jpickup** [comment]
First off, @Abasz, thanks for all your analysis here.

@surneym, I've created an updated version of the app that creates unqiue serial numbers on workouts. It's on branch feature/serial_no. Please can you try this? The output works on my watch, however so did the files where serial number was always set to 1 so that isn't much of a test!

Also, please note what @Abasz has said above, namely that files should be dropped into the NewFiles directory on the watch rather than the workouts directory. Also, there are definitely limits on how many future workouts can be scheduled and possibly how far into the future. These are Garmin limitations, not the app. So please also try setting up a schedule that starts in the near future rather than next year.

Please let me know if this helps.

Thanks.

**Comment by surbeym** [comment]
> First off, @Abasz, thanks for all your analysis here.
> 
> @surbeym, I've created an updated version of the app that creates unique serial numbers on workouts. It's on branch feature/serial_no. Please can you try this? The output works on my watch, however so did the files where serial number was always set to 1 so that isn't much of a test!
>
> Please let me know if this helps.
> 
> Thanks.
> 
> [Installer link on branch](https://github.com/jpickup/GarminTools/tree/feature/serial_no/installer)

Hello,

Great news reran the xls through the new version, and copied the files (Schedule and workouts, all 111 files) over to the NewFiles directory, did get the warning about too many, as I expected along with you guys. 


I started the excel sheet at Nov 5, 2023 and goes to March 9, 2023 and the watch is show all up to Dec 5, 2023. This is also expected behavior!


I shall dedicate my mile 20 on my first marathon to you. Willo be thinking of the support you gave when/if I hit the wall.

![IMG_20231104_101659](https://github.com/jpickup/GarminTools/assets/2860120/303e2388-e924-4553-a315-a054c7c89204)

**Comment by jpickup** [comment]
Glad to hear that it's all working now and I hope the marathon training goes well. Are you on Strava? If so I'd like to follow you if that's OK. Here's a link to my [Strava profile](https://www.strava.com/athletes/3348955).

**Comment by surbeym** [comment]
Hey I don't use Strava currently, as I keep my health data to only Garmin. Here is a link to my Garmin Profile https://connect.garmin.com/modern/profile/b5ec5017-f22a-4546-82df-db8d89e0d0c8
